{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "626896f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import yaml\n",
    "import json\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dff2361c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwl_dir = '/home/estorrs/pecgs-fusion/cwl/'\n",
    "Path(cwl_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c22d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = m['id']\n",
    "fq_1 = m['rna-seq']['tumor']['R1']['filepath']\n",
    "fq_2 = m['rna-seq']['tumor']['R2']['filepath']\n",
    "cpu = 16\n",
    "genome_lib_dir = '/data/pecgs/fusion/STAR-Fusion_dependencies/GRCh38_gencode_v37_CTAT_lib_Mar012021.plug-n-play/ctat_genome_lib_build_dir/'\n",
    "genome_db = '/data/pecgs/fusion/ericscript_dependencies/ericscript_db_homosapiens_ensembl84/'\n",
    "# ericscript = '/data/pecgs/fusion/ericscript-0.5.5/ericscript.pl'\n",
    "bwts = '/data/pecgs/fusion/Integrate_dependencies/bwts/'\n",
    "integrate_fasta = '/data/pecgs/fusion/Integrate_dependencies/STAR/hg38.fa'\n",
    "integrate_annotations = '/data/pecgs/fusion/Integrate_dependencies/annot.ensembl.GRCh38.txt'\n",
    "combine_call_script = '/data/pecgs/fusion/Fusion_hg38_scripts/combine_call.pl'\n",
    "filter_script = '/data/pecgs/fusion/Fusion_hg38_scripts/filter.pl'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c60f914d",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = {\n",
    "    'fq_1': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'fq_2': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'genome_lib_dir': {\n",
    "        'class': 'Directory',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'genome_db': {\n",
    "        'class': 'Directory',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'bwts': {\n",
    "        'class': 'Directory',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'integrate_fasta': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'integrate_annotations': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'sample': 'a_string',\n",
    "    'cpu': 'a_int',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3efaf6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(template, open(os.path.join(cwl_dir, 'template.fusion.yaml'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "20e0d4d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwl = {\n",
    "    'class': 'CommandLineTool',\n",
    "    'cwlVersion': 'v1.0',\n",
    "    '$namespaces': {'sbg': 'https://www.sevenbridges.com/'},\n",
    "    'id': 'fusion',\n",
    "    'label': 'fusion',\n",
    "    'baseCommand': ['/pecgs-fusion/fusion/fusion.py'],\n",
    "    'inputs': [\n",
    "        {\n",
    "            'id': 'sample',\n",
    "            'type': 'string',\n",
    "            'inputBinding': {\n",
    "                'position': '1'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'fq_1',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'position': '2'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'fq2',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'position': '3'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'genome_lib_dir',\n",
    "            'type': 'Directory',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--genome-lib-dir',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'genome_db',\n",
    "            'type': 'Directory',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--genome-db',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'bwts',\n",
    "            'type': 'Directory',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--bwts',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'integrate_fasta',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--integrate-fasta',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'integrate_annotations',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--integrate-annotations',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'cpu',\n",
    "            'type': 'string?',\n",
    "            'inputBinding': {\n",
    "                'prefix': '-cpu',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "    ],\n",
    "    'arguments': [\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--combine-call-script',\n",
    "            'valueFrom': '/pecgs-fusion/fusion/combine_call.pl'\n",
    "        },\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--filter-script',\n",
    "            'valueFrom': '/pecgs-fusion/fusion/filter.pl'\n",
    "        }\n",
    "    ],\n",
    "    'outputs': [\n",
    "        {\n",
    "            'id': 'output_summary',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'output'},     \n",
    "        },\n",
    "        {\n",
    "            'id': 'output_dis',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'output_dis'},     \n",
    "        },\n",
    "        {\n",
    "            'id': 'output_germline',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'output_germline'},     \n",
    "        },\n",
    "        {\n",
    "            'id': 'output_somatic',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'output_somatic'},     \n",
    "        }\n",
    "    ],\n",
    "    'requirements': [\n",
    "        {\n",
    "            'class': 'DockerRequirement',\n",
    "            'dockerPull': 'estorrs/pecgs_fusion:0.0.1'\n",
    "        },\n",
    "        {\n",
    "            'class': 'ResourceRequirement',\n",
    "            'ramMin': 50000\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7023448",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(cwl, open(os.path.join(cwl_dir, 'fusion.cwl'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad4ea05",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
